src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs

```diff
@@ -104,6 +117,63 @@ public class BuildIntegratedProjectAction : NuGetProjectAction
             Sources = sources;
             OriginalActions = originalActions;
             InstallationContext = installationContext;
+            ActionAndContextList = originalActions.Select(e => (e, installationContext)).ToList();
+        }
+
+        internal BuildIntegratedProjectAction(NuGetProject project,
+            PackageIdentity packageIdentity,
+            NuGetProjectActionType nuGetProjectActionType,
+            LockFile originalLockFile,
+            RestoreResultPair restoreResultPair,
+            IReadOnlyList<SourceRepository> sources,
+            IReadOnlyList<(NuGetProjectAction, BuildIntegratedInstallationContext)> originalActionsAndInstallationContexts,
+            VersionRange versionRange)
+            : base(packageIdentity, nuGetProjectActionType, project, sourceRepository: null, versionRange)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            if (packageIdentity == null)
+            {
+                throw new ArgumentNullException(nameof(packageIdentity));
+            }
+
+            if (originalLockFile == null)
+            {
+                throw new ArgumentNullException(nameof(originalLockFile));
+            }
+
+            if (restoreResultPair == null)
+            {
+                throw new ArgumentNullException(nameof(restoreResultPair));
+            }
+
+            if (sources == null)
+            {
+                throw new ArgumentNullException(nameof(sources));
+            }
+
+            if (originalActionsAndInstallationContexts == null)
+            {
+                throw new ArgumentNullException(nameof(originalActionsAndInstallationContexts));
+            }
+
+            if (originalActionsAndInstallationContexts.Count < 1)
+            {
+                throw new ArgumentException("Must contain at least 1 element.", nameof(originalActionsAndInstallationContexts));
+            }
+
+            OriginalLockFile = originalLockFile;
+            RestoreResult = restoreResultPair.Result;
+            RestoreResultPair = restoreResultPair;
+            Sources = sources;
+            ActionAndContextList = originalActionsAndInstallationContexts;
+#pragma warning disable CS0618 // Type or member is obsolete
+            OriginalActions = originalActionsAndInstallationContexts.Select(e => e.Item1).ToList();
+            InstallationContext = originalActionsAndInstallationContexts[0].Item2;
+#pragma warning restore CS0618 // Type or member is obsolete
```

Comment:
I'm really curious to understand why we should access the obsolete members here. We already assigned the new property, namely ActionAndContextList, with the correct value in this constructor, which isn't obsolete. And have we done the same in the constructor that is marked as obsolete in this PR.

Reply:
The constructor is new, there's a new property, but that  doesn't mean that the old contract should be broken. 
Normally, if this wasn't a public API, I'd remove the old properties, but that's not possible without a breaking change currently. 
So until then, we make sure that the type is correctly populated.

Reply:
The two properties `OriginalActions` and `InstallationContext` are currently unused in the product code. If a caller uses the new constructor, we'll assign a value to the newly introduced property in this PR, `ActionAndContextList`, which the product code references. However, if the caller uses the old constructor, we'll generate the `ActionAndContextList` property using `OriginalActions` and `InstallationContext`. This new property will then be used throughout the rest of the product code. Therefore, I'm uncertain about the necessity of assigning values to the old or obsolete properties in the new constructor.

Reply:
We don't know who/how people use our code unfortunately.
If people say we're not worried about that, I can remove it, but I was erring on the cautious side.